### ISSUE TITLE
[IMPROVEMENT] Selective memory addition gate with history-based utility deletion (Experience-Following Property)

### ISSUE THREAD

**Issue by Lexus2016** [Lexus2016/hermes-agent-evolution#1270]
### Problem Statement
The evolution pipeline appends every research report, every issues file, every implementation log indiscriminately — an **add-all** memory policy. The ACL 2026 memory-management paper (aclanthology.org/2026.acl-long.27) proves this self-degrades via three mechanisms:

1. **Experience-Following Property**: a high input-similarity between the current task and a retrieved memory record yields a high output-similarity (Pearson r ≈ 1.0). Agents *imitate* retrieved demonstrations — this is how memory helps AND how it harms. A retrieved noisy/misaligned record is imitated and amplifies the error.
2. **Error Propagation**: if a retrieved memory record contains noisy/incorrect outputs, the agent replicates and amplifies the error; if that execution is re-added to memory, the error propagates. This is the memory analogue of the BenchJack evaluator-gaming finding.
3. **Misaligned Experience Replay**: some records that pass the evaluator's quality filter still consistently lead to poor downstream execution because they are misaligned with the current task distribution.

Quantitatively, **add-all is worse than fixed-memory** on 3 of 4 agents tested; **strict-evaluator + selective addition beats add-all by 22–25 points**; and **history-based deletion** (remove a record retrieved ≥n times if its average downstream utility is below threshold) *improves* performance beyond no-deletion on real agents. A key result: a small fine-tuned quality evaluator (300 training examples) recovers most of an oracle evaluator's gain, beating a large generic LLM judge — "directly applying a vanilla LLM as a trajectory evaluator may lead to a more severe negative impact than manually crafting a small but high-quality dataset."

This is the mechanism behind the SkillHone "repeating a previously successful repair that no longer matches" failure (2026-07-24): the agent retrieves a similar-looking past decision and imitates it, even if the current failure mode differs.

### Proposed Solution
Replace add-all with a selective memory policy, applied to the pipeline's own artifacts (research reports, issues, implementation logs) and as a recommendation for the agent's memory/skills:

1. **Selective-addition gate**: only store an artifact if a quality evaluator judges it high-quality. Start with a small specialized quality classifier (the paper shows 300 training examples suffice) rather than a generic GPT-judge-everything — the generic LLM judge can be net-negative.
2. **History-based deletion**: delete memories/skills by **downstream utility**, not age. Track how often each memory/skill is retrieved and whether the retrieval correlated with a successful outcome; delete low-utility records. This is a richer signal than periodic (age-based) deletion or the SWE-Pruner token counts.
3. **Misaligned-record detection**: flag records that pass the quality filter but consistently lead to poor downstream execution; quarantine or delete them.
4. **Error-propagation guard**: never re-add an execution that imitated a noisy record back into memory (the amplification loop).

### Value Proposition
- **Impact**: High (0.8) — add-all provably self-degrades; selective addition + history-based deletion provably improves.
- **Effort**: Low-Medium (0.2) — utility tracking is a structured-logging change on top of existing retrieval; the quality gate is a classifier on 300 examples.
- **Priority Score**: 1.44

### Research Evidence
- [Experience-Following Property — ACL 2026 long paper #27](https://aclanthology.org/2026.acl-long.27.pdf) (Harvard, U. Georgia, Michigan State, U. Minnesota)
- Code: https://github.com/yuplin2333/agent_memory_manage
- In-repo related: SkillHone #1256 (decision history), memory plugins (mem0, holographic, hindsight)

### Implementation Plan
1. Add a retrieval-utility log: each time a memory/skill is retrieved, record (record_id, retrieval_context, downstream outcome).
2. Add a history-based-deletion function: remove records retrieved ≥n times with average downstream utility below threshold.
3. Build a small quality classifier for pipeline artifacts (research reports, implementation logs) — 300 positive/negative examples.
4. Wire the selective-addition gate into the artifact-store step (research save, issues save, implementation log save).
5. Add a misaligned-record detector: flag records with high retrieval count but low downstream-outcome correlation.

### Success Criteria
- [ ] No artifact is stored without passing the quality gate
- [ ] History-based deletion removes low-utility records without performance regression
- [ ] Retrieval-utility log records downstream outcome per retrieval
- [ ] Misaligned-record detector flags at least one record class in a sample run

**Comment by Lexus2016** [comment]
## Rework brief — PR #1275 closed (dead code)

The PR landed the memory gate **module** (`scripts/evolution_memory_gate.py`) but did not wire it into the pipeline. The module's `evaluate(payload)` is called by nothing — a gate that nothing invokes is not a gate.

### What's missing
1. **Call site.** No pipeline script calls the memory gate. It should be invoked when the pipeline is deciding whether to store an artifact (research report, skill, etc.) in memory/skills.
2. **SKILL.md / cron integration.** No reference in any cron YAML or SKILL.md.
3. **Integration test.** `test_evolution_memory_gate_wiring.py` tests `evaluate()` in isolation, not pipeline integration.

### Rework instructions
1. Wire the gate into the pipeline at the point where artifacts are considered for memory/skill storage (likely in `evolution_orchestrator.py` or a post-evaluation stage).
2. Update the relevant cron YAML / SKILL.md.
3. Rewrite the test to verify the gate is actually invoked from its call site.
4. Re-open as a new PR.

**Comment by Lexus2016** [comment]
Implementation cycle 2026-07-26 — status update (no code change this cycle).

Both prior attempts closed by the owner:
- PR #1275: dead code.
- PR #1280: closed as **incoherent** — "runs after artifacts have already been written, so it only reports decisions and never admits, quarantines, deletes, or blocks memory records. Its historical inputs are permanently empty, and it treats impact as quality."

Owner's rework brief: the underlying need is valid, but it needs a future implementation at the **actual artifact-write boundaries** with:
1. Quality evidence at write time.
2. Lineage tracking.
3. Quarantine storage.
4. An explicit fail-open contract.

This requires intercepting the artifact-store step at the real write boundary (not post-hoc reporting), which is an architectural change to where/when the gate runs. Keeping **open + blocked** rather than forcing a third post-write evaluate() attempt. Same pattern as #1267 and #1268.

**Comment by Lexus2016** [comment]
Blocked by #1363 — the missing input is now filed as its own issue.

The `blocked` label on this issue had no blocker recorded. Verified today against `main`: `evolution_trajectory_logger.py` is only ever called by `evolution_funnel.py` (it records `tool='evolution_funnel'` — the pipeline logging itself), and `agent.save_trajectories` defaults to `False` with nothing in the pipeline reading its output. There is no capture of real agent tool calls anywhere.

That is why both attempts here were closed as incoherent: with no real data to read, each implementation had to synthesize a substitute, and the substitute is what the review correctly rejected. The same pattern hit #1267, #1268 and #1270 — six closed PRs on one missing input.

Recommend not re-attempting this until #1363 lands. Re-attempting against synthetic trajectories will produce the same outcome a third time.

**Comment by Lexus2016** [comment]
**Unblocked** — #1363 merged as `2f631a761`. The input this issue was waiting on now exists.

## What is available

Real agent tool calls are captured at `finalize_turn` and written to `~/.hermes/evolution/trajectories/<date>_<session>.jsonl`, one JSON object per turn:

```json
{
  "date": "2026-07-28",
  "session_id": "...",
  "completed": true,
  "task_key": "f70300a2aca1337c",
  "entries": [
    {"tool": "read_file", "args_summary": {...}, "result_status": "success",
     "result_summary": "...", "timestamp": "...", "duration_ms": null}
  ]
}
```

Read it with `evolution_trajectory_logger.load_trajectory` (single object) or line-by-line for a whole session.

## Three things to know before implementing

**1. Capture is opt-in.** It is behind `HERMES_EVOLUTION_CAPTURE`, default OFF. Redaction catches credential-shaped keys, not sensitive values in ordinary fields, so it is not something to run on every interactive session unasked. **The cron stage that needs the data must set it.** If you implement against an empty directory, that is why.

**2. `completed` distinguishes three states, not two.** `true`/`false` are recorded outcomes; **absent means "not recorded"** — the pre-#1363 cron-stage trajectories have no such field. Treating absent as failure would poison any outcome-conditioned analysis.

**3. `task_key` is a salted HMAC, not a hash of the prompt.** It pairs runs of the same task *within one store*. It is deliberately not comparable across machines, and the descriptor cannot be recovered from it.

## What is still missing, honestly

- **No per-call latency.** `duration_ms` exists in the schema but nothing populates it yet — the capture runs after the turn, where per-call timing is no longer available. #1442's 3-timestamp framework needs this and should treat it as its own first slice.
- **Only turns that made tool calls are recorded.** A turn with no calls writes nothing, so the store is not a complete session history.
- **A call with no result is recorded as `pending`.** Deliberate — a call that never returned is signal for #1268's error-recovery dimension — but it is a third status your classifier must handle alongside `success`/`failure`.

## Verify before building

The record was designed against the stated needs of #1268, #1359 and #1436, but no consumer has been implemented against it yet. Check it actually carries what this issue needs before writing the implementation, and say so on #1363 if it does not — a second round on the record shape is much cheaper than a seventh rejected PR.

**Comment by Lexus2016** [comment]
Decomposed into child issues:
- #1480 — Retrieval-utility logging + history-based deletion (Child A, first implementable slice, ~150 lines)

Steps 3-5 (quality classifier, selective-addition gate wiring, misaligned-record detector) are deferred to later child issues pending utility data accumulation.

**Comment by Lexus2016** [comment]
## Accepted, blocked on decomposition (needs-split)

The Experience-Following Property is a real failure mode for a fork whose memory grows monotonically, and a utility-based deletion gate is the right general shape.

Two independent halves that should not land together:
1. **The addition gate** — deciding what is worth storing. Testable on its own.
2. **History-based utility deletion** — deleting stored memory. Higher risk: a wrong utility function silently destroys context, and unlike a bad addition gate the damage is not recoverable.

Do the addition gate first, and only build deletion once there is a measured utility signal to base it on. Deleting on an unvalidated heuristic is worse than keeping too much.
